kbs/src/jwe.rs

```diff
@@ -43,23 +121,277 @@ pub fn jwe(tee_pub_key: TeePubKey, payload_data: Vec<u8>) -> Result<Response> {
+    let z = encrypter_secret
+        .diffie_hellman(&public_key)
+        .raw_secret_bytes()
+        .to_vec();
+    let mut key_derivation_materials = Vec::new();
+    key_derivation_materials.extend_from_slice(&(ECDH_ES_A256KW.len() as u32).to_be_bytes());
+    key_derivation_materials.extend_from_slice(ECDH_ES_A256KW.as_bytes());
+    key_derivation_materials.extend_from_slice(&(0_u32).to_be_bytes());
+    key_derivation_materials.extend_from_slice(&(0_u32).to_be_bytes());
+    key_derivation_materials.extend_from_slice(&AES_GCM_256_KEY_BITS.to_be_bytes());
+    let mut wrapping_key = vec![0; 32];
+    concat_kdf::derive_key_into::<rsa::sha2::Sha256>(
+        &z,
+        &key_derivation_materials,
+        &mut wrapping_key,
+    )
+    .map_err(|e| anyhow!("failed to do concat KDF: {e:?}"))?;
+    let wrapping_key: [u8; 32] = wrapping_key
+        .try_into()
+        .map_err(|_| anyhow!("invalid bytes length of AES wrapping key"))?;
+    let wrapping_key: KekAes256 = Kek::new(&GenericArray::from(wrapping_key));
+    let mut encrypted_key = vec![0; 40];
+    encrypted_key.resize(40, 0);
+    let cek = cek.to_vec();
+    wrapping_key
+        .wrap(&cek, &mut encrypted_key)
+        .map_err(|e| anyhow!("failed to do AES wrapping: {e:?}"))?;
+
+    let point = EncodedPoint::from(encrypter_secret.public_key());
+    let epk_x = point
+        .x()
+        .ok_or(anyhow!("invalid public key: without coordicate X"))?;
+    let epk_y = point
+        .y()
+        .ok_or(anyhow!("invalid public key: without coordicate Y"))?;
+    let epk_x = URL_SAFE_NO_PAD.encode(epk_x);
+    let epk_y = URL_SAFE_NO_PAD.encode(epk_y);
+    let protected = ProtectedHeader {
+        alg: ECDH_ES_A256KW.to_string(),
+        enc: AES_GCM_256_ALGORITHM.to_string(),
+        other_fields: json!({
+            "epk": {
+                "crv": P256_CURVE,
+                "kty": EC_KTY,
+                "x": epk_x,
+                "y": epk_y
+            }
+        })
+        .as_object()
+        .unwrap()
+        .clone(),
+    };
+
+    // 3. Encrypt content with CEK
+    let mut cek_cipher = Aes256Gcm::new(GenericArray::from_slice(&cek));
+
+    let iv = rand::thread_rng().gen::<[u8; 12]>();
+    let nonce = Nonce::from_slice(&iv);
+    let aad = protected.generate_aad().context("Generate JWE AAD")?;
+
+    let tag = cek_cipher
+        .encrypt_in_place_detached(nonce, &aad, &mut payload_data)
+        .map_err(|e| anyhow!("AES encrypt Resource payload failed: {e}"))?;
 
     Ok(Response {
-        protected: serde_json::to_string(&protected_header)
-            .context("serde protected_header failed")?,
-        encrypted_key: URL_SAFE_NO_PAD.encode(wrapped_sym_key),
-        iv: URL_SAFE_NO_PAD.encode(iv),
-        ciphertext: URL_SAFE_NO_PAD.encode(encrypted_payload_data),
-        tag: "".to_string(),
+        protected,
+        encrypted_key,
+        iv: iv.into(),
+        ciphertext: payload_data,
+        aad: None,
```

Comment:
set aad to None is deliberated? why?

Reply:
oh, checking the client side code in https://github.com/confidential-containers/guest-components/pull/820/files `let aad = response.protected.generate_aad()?;`, I guess aad is generated but not passed. Got.